src/hotspot/share/opto/loopnode.cpp

```diff
@@ -2309,97 +2198,313 @@ bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_
+    return false;
+  }
+
+  // Variables needed by convert() to do the actual conversion. We set these fields iff a counted loop is confirmed.
+  // This way the is_counted_loop() can run multiple times with possibly different structures.
+  _structure = structure;
+
+  _stride_con = stride_con;
+  _final_correction = final_correction;
+  _includes_limit = includes_limit;
+
+  _insert_stride_overflow_limit_check = insert_stride_overflow_limit_check;
+  _insert_init_trip_limit_check = insert_init_trip_limit_check;
+
+#ifdef ASSERT
+  _checked_for_counted_loop = true;
+#endif
+
+  return true;
+}
+
+bool CountedLoopConverter::build_loop_structure(CountedLoopConverter::LoopStructure& structure) {
+  PhaseIterGVN* igvn = &_phase->igvn();
+
+  Node* back_control = _phase->loop_exit_control(_head, _loop);
+  if (back_control == nullptr) {
+    return false;
+  }
+
+  PhaseIdealLoop::LoopExitTest exit_test = _phase->loop_exit_test(back_control, _loop);
+  if (exit_test.cmp == nullptr || exit_test.cmp->Opcode() != Op_Cmp(_iv_bt)) {
+    return false; // Avoid pointer & float & 64-bit compares
+  }
+
+  Node* incr = exit_test.incr;
+  if (exit_test.incr->Opcode() == Op_Cast(_iv_bt)) {
+    incr = incr->in(1);
+  }
+
+  const PhaseIdealLoop::LoopIVIncr iv_incr = PhaseIdealLoop::loop_iv_incr(incr, _head, _loop);
+  if (iv_incr.incr == nullptr) {
+    return false;
+  }
+
+  CountedLoopNode::TruncatedIncrement
+      increment = CountedLoopNode::match_incr_with_optional_truncation(iv_incr.incr, _iv_bt);
+  if (increment.incr == nullptr) {
+    return false; // Funny increment opcode
+  }
+  assert(increment.incr->Opcode() == Op_Add(_iv_bt), "wrong increment code");
+
+  const PhaseIdealLoop::LoopIvStride stride = PhaseIdealLoop::loop_iv_stride(increment.incr);
+  if (stride.stride == nullptr) {
+    return false;
+  }
+
+  Node* xphi = stride.xphi;
+  // Iteratively uncast the loop induction variable
+  // until no more CastII/CastLL nodes are found.
+  while (xphi->Opcode() == Op_Cast(_iv_bt)) {
+    xphi = xphi->in(1);
+  }
+
+  PhiNode* phi = PhaseIdealLoop::loop_iv_phi(xphi, iv_incr.phi_incr, _head);
+  if (phi == nullptr ||
+      (increment.trunc1 == nullptr && phi->in(LoopNode::LoopBackControl) != increment.incr) ||
+      (increment.trunc1 != nullptr && phi->in(LoopNode::LoopBackControl) != increment.trunc1)) {
+    return false;
+  }
+
+  SafePointNode* sfpt = find_safepoint(back_control);
+
+  structure.back_control = back_control;
+
+  structure.cmp = exit_test.cmp;
+  structure.incr = incr;
+  structure.limit = exit_test.limit;
+  structure.mask = exit_test.mask;
+  structure.cl_prob = exit_test.cl_prob;
+
+  structure.iv_incr = iv_incr.incr;
+  structure.phi_incr = iv_incr.phi_incr;
+
+  structure.trunc_incr = increment.incr;
+  structure.trunc1 = increment.trunc1;
+  structure.trunc2 = increment.trunc2;
+  structure.trunc_type = increment.trunc_type;
+
+  structure.stride = stride.stride;
+
+  structure.phi = phi;
+
+  structure.sfpt = sfpt;
```

Comment:
Are you later really going to use all these fields? I haven't double-checked. Another note here: I think it would be better to hide the fields and provide accessor methods. Otherwise, everyone can update them.